src/vizier/services/metadata/controllers/tracepoint/tracepoint.go

```diff
@@ -295,26 +297,166 @@ func (m *Manager) UpdateAgentTracepointStatus(tracepointID *uuidpb.UUID, agentID
+// that match those selectors. For that list of agents, we send out a tracepoint request with
+// a new tracepointDeployment. If multiple programs may have the same list of allowed agents,
+// we collapse them into one tracepoint deployment and send those in one request.
+// Note: stirling current only supports one tracepoint per tracepoint deployment.
+func (m *Manager) RegisterTracepoint(agents []*agentpb.Agent, tracepointID uuid.UUID, tracepointDeployment *logicalpb.TracepointDeployment) error {
+	// Map where key is the hash of agent IDs and value is the list of programs for those agents.
+	agentsHashToPrograms := make(map[string][]*logicalpb.TracepointDeployment_TracepointProgram)
+	// Map where key is the hash of agent IDs and value is a struct containing the list of agents and their IDs.
+	agentsHashToAgents := make(map[string]struct {
+		Agents   []*agentpb.Agent
+		AgentIDs []uuid.UUID
+	})
+
+	for _, prgm := range tracepointDeployment.Programs {
+		validAgents := agents // Start with all agents as potential targets.
+
+		for _, selector := range prgm.Selectors {
+			validAgents = m.FilterAgentsBySelector(validAgents, selector)
+		}
+
+		// Generate a hash for the list of valid agents.
+		agentIDs := make([]uuid.UUID, len(validAgents))
+		for i, agt := range validAgents {
+			agentIDs[i] = utils.UUIDFromProtoOrNil(agt.Info.AgentID)
+		}
+		hash := utils.HashUUIDs(agentIDs)
+
+		agentsHashToPrograms[hash] = append(agentsHashToPrograms[hash], prgm)
+		agentsHashToAgents[hash] = struct {
+			Agents   []*agentpb.Agent
+			AgentIDs []uuid.UUID
+		}{Agents: validAgents, AgentIDs: agentIDs}
+	}
+
+	for hash, validAgentsForProgram := range agentsHashToAgents {
+		// Build a new TracepointDeployment with the group of programs.
+		newDeployment := &logicalpb.TracepointDeployment{
+			Name:           tracepointDeployment.Name,
+			TTL:            tracepointDeployment.TTL,
+			DeploymentSpec: tracepointDeployment.DeploymentSpec,
+			Programs:       agentsHashToPrograms[hash],
+		}
+
+		// Send a RegisterTracepointRequest to each agent that supports this program.
+		tracepointReq := messagespb.VizierMessage{
+			Msg: &messagespb.VizierMessage_TracepointMessage{
+				TracepointMessage: &messagespb.TracepointMessage{
+					Msg: &messagespb.TracepointMessage_RegisterTracepointRequest{
+						RegisterTracepointRequest: &messagespb.RegisterTracepointRequest{
+							TracepointDeployment: newDeployment,
+							ID:                   utils.ProtoFromUUID(tracepointID),
+						},
 					},
 				},
 			},
-		},
-	}
-	msg, err := tracepointReq.Marshal()
-	if err != nil {
-		return err
+		}
+		msg, err := tracepointReq.Marshal()
+		if err != nil {
+			return err
+		}
+
+		agentIDs := validAgentsForProgram.AgentIDs
+		for i, agt := range validAgentsForProgram.Agents {
+			agentIDs[i] = utils.UUIDFromProtoOrNil(agt.Info.AgentID)
+		}
+
+		err = m.agtMgr.MessageAgents(agentIDs, msg)
```

Comment:
Isn't this for loop redundant now? I was hoping we would be able to remove it entirely and write the following:

```go
err = m.agtMgr.MessageAgents(validAgentsForProgram.AgentIDs, msg)
```

Reply:
Yes sorry, I forgot to remove that loop. Should be fixed now.